third-party-programs/Velocity-Bench/bitcracker/README.md

```diff
@@ -1,85 +1,175 @@
+$ dpct --in-root=. -p=./build/compile_commands.json --out-root=out --gen-build-script --cuda-include-path=/usr/local/cuda/include
 ```
-Note:
-- We use the flag `-DUSE_AMDHIP_BACKEND=gfx90a` for MI250. Use the correct value for your GPU.
+Description of the options:
+ * `--in-root`: provide input files to specify where to locate the CUDA files that needs migration.
+ * `-p`: specify compilation database to migrate the whole project.
+ * `--out-root`: designate where to generate the resulting files (default is `dpct_output`).
+ * `--gen-build-script`: generate the `Makefile.dpct` for the migrated code.
 
-For NVIDIA GPU -  
-First source clang++ compiler. Then,
+Now you can see the migrated files in the `out` folder as follow:
 ```
-cd bitcracker/SYCL
-mkdir build
-cd build
-CXX=clang++ cmake -DUSE_NVIDIA_BACKEND=YES -DUSE_SM=80 ..
-make -sj
+   out/
+   ├── MainSourceFiles.yaml
+   ├── Makefile.dpct
+   └── src
+      ├── aes.h
+      ├── aes.h.yaml
+      ├── attack.dp.cpp
+      ├── bitcracker.h
+      ├── bitcracker.h.yaml
+      ├── main.dp.cpp
+      ├── sha256.h
+      ├── sha256.h.yaml
+      ├── utils.dp.cpp
+      └── w_blocks.dp.cpp
 ```
-Note:
-- We use the flag `-DUSE_SM=80` for A100 or `-DUSE_SM=90` for H100.
+### 4 Review the migrated source code and fix all `DPCT` warnings
 
-## To build for CUDA
+SYCLomatic and [Intel® DPC++ Compatibility Tool](https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compatibility-tool.html) define a list of `DPCT` warnings and embed the warning in migrated source code if need manual effort to check. All the warnings in the migrated code should be reviewed and fixed. For detail of `DPCT` warnings and corresponding fix examples, refer to [Intel® DPC++ Compatibility Tool Developer Guide and Reference](https://www.intel.com/content/www/us/en/develop/documentation/intel-dpcpp-compatibility-tool-user-guide/top/diagnostics-reference.html) or [SYCLomatic doc page](https://oneapi-src.github.io/SYCLomatic/dev_guide/diagnostics-reference.html).
 
+Fix the warning in migrated bitcracker code:
 ```
-cd bitcracker/CUDA
-mkdir build
-cd build
-CXX=nvcc cmake -DUSE_SM=80 ..
-make -sj
+$ cat ${bitcracker_HOME}/CUDA/out/src/attack.dp.cpp
+...
+/*
+DPCT1110:3: The total declared local variable size in device function
+decrypt_vmk_with_mac exceeds 128 bytes and may cause high register pressure.
+Consult with your hardware vendor to find the total register size available and
+adjust the code, or use smaller sub-group size to avoid high register pressure.
+*/
+
+void decrypt_vmk_with_mac(
+...
 ```
+This message is shown because the Compatibility Tool finding the user-declared private memeory size of local variable in the kernel may exceed 128 bytes which is the largest register size for each work-item. It may cause high register pressure. For more details, you can refer [oneAPI GPU Optimization Guide](https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2024-0/overview.html)
```

Comment:
```suggestion
This message is shown because the Compatibility Tool finding the user-declared private memory size of the local variable in the kernel may exceed 128 bytes, which is the largest register size for each work-item. It may cause high register pressure. For more details, you can refer [oneAPI GPU Optimization Guide](https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2024-0/overview.html)
```